File: fft-ts/src/fft47.ts
import { minus, plus, times, rot90, Complex, timesScalar } from "complex/dst/Complex";
import { ComplexArray, getComplex, makeComplexArray, setComplex } from "complex/dst/ComplexArray";
import { fallbackFFT } from "./fallbackFFT";
import { FFTPrep } from "./fft_types";


const TAU = 2 * Math.PI; // https://en.wikipedia.org/wiki/Turn_(angle)#Tau_proposals

export const fft_prepare: FFTPrep = n => {
  const cosines = new Float64Array(n);
  for (let i = 0; i < n; i++) {
    cosines[i] = Math.cos(TAU * i / n);
  }

  const quarterN = n >>> 2;
  const permute = new Uint32Array(quarterN);
  for (let len = quarterN, fStride = 1; len > 1; len >>>= 1, fStride <<= 1) {
    const halfLen = len >>> 1;
    for (let out_offset = 0; out_offset < quarterN; out_offset += len) {
      const limit = out_offset + len;
      for (let out_offset_odd = out_offset + halfLen; out_offset_odd < limit; out_offset_odd++) {
        permute[out_offset_odd] += fStride;
      }
    }
  }

  return function fft(f: ComplexArray, out: ComplexArray, direction: number = 1): void {
    if (n <= 2) return fallbackFFT(n, f, out);

    const nMask = n - 1;
    const negDirection = -direction;

    const rotation = (x: number): Complex => ({
      re: cosines[x & nMask],
      im: cosines[(quarterN - x) & nMask]
    })

    for (let out_offset = 0; out_offset < n;) {
      let offset = permute[out_offset >> 2];
      const b0 = getComplex(f, offset); offset += quarterN;
      const b2 = getComplex(f, offset); offset += quarterN;
      const b1 = getComplex(f, offset); offset += quarterN;
      const b3 = getComplex(f, offset);

      const c0 =                   plus (b0, b1);
      const c1 =                   minus(b0, b1);
      const c2 =                   plus (b2, b3);
      const c3 = timesScalar(rot90(minus(b2, b3)), negDirection);

      setComplex(out, out_offset++, plus (c0, c2));
      setComplex(out, out_offset++, plus (c1, c3));
      setComplex(out, out_offset++, minus(c0, c2));
      setComplex(out, out_offset++, minus(c1, c3));
    }

    let len = 8, rStride = direction * (n >>> 3);
    for (; len < n; len <<= 2, rStride >>= 2) {
      const halfLen = len >> 1;
      {
        for (let out_offset = 0; out_offset < n;) {
          const i0 = out_offset; out_offset += halfLen;
          const i1 = out_offset; out_offset += halfLen;
          const i2 = out_offset; out_offset += halfLen;
          const i3 = out_offset; out_offset += halfLen;
  
          const b0 = getComplex(out, i0);
          const b1 = getComplex(out, i1);
          const b2 = getComplex(out, i2);
          const b3 = getComplex(out, i3);
  
          const c0 =                   plus (b0, b1);
          const c1 =                   minus(b0, b1);
          const c2 =                   plus (b2, b3);
          const c3 = timesScalar(rot90(minus(b2, b3)), negDirection);
  
          setComplex(out, i0, plus (c0, c2));
          setComplex(out, i1, plus (c1, c3));
          setComplex(out, i2, minus(c0, c2));
          setComplex(out, i3, minus(c1, c3));
        }
      }
      const rStride1 = rStride >> 1;
      const rStride2 = rStride;
      const rStride3 = rStride2 + rStride1;
      let rOffset1 = -rStride1;
      let rOffset2 = -rStride2;
      let rOffset3 = -rStride3;
      for (let k = 1; k < halfLen; k++) {
        // TODO Some bit fiddling with rOffset[123] to restrict cosine lookups
        // to the first quadrant?  Then shorten the cosines array.
        const r1 = rotation(rOffset1); rOffset1 -= rStride1;
        const r2 = rotation(rOffset2); rOffset2 -= rStride2;
        const r3 = rotation(rOffset3); rOffset3 -= rStride3;
        for (let out_offset = k; out_offset < n;) {
          const i0 = out_offset; out_offset += halfLen;
          const i1 = out_offset; out_offset += halfLen;
          const i2 = out_offset; out_offset += halfLen;
          const i3 = out_offset; out_offset += halfLen;

          const b0 =       getComplex(out, i0)     ;
          const b1 = times(getComplex(out, i1), r2);
          const b2 = times(getComplex(out, i2), r1);
          const b3 = times(getComplex(out, i3), r3);

          const c0 =                   plus (b0, b1);
          const c1 =                   minus(b0, b1);
          const c2 =                   plus (b2, b3);
          const c3 = timesScalar(rot90(minus(b2, b3)), negDirection);

          setComplex(out, i0, plus (c0, c2));
          setComplex(out, i1, plus (c1, c3));
          setComplex(out, i2, minus(c0, c2));
          setComplex(out, i3, minus(c1, c3));
        }
      }
    }
    if (len === n) {
      // If we come here, n is not a power of 4 (but still a power of 2).
      // So we need to run one extra round of 2-way butterflies.
      const halfLen = len >>> 1;

      // TODO Roll this back into the following loop?
      // Saving a single complex multiplicatin is probably not worth the extra code.
      {
        const z0 = getComplex(out, 0);
        const z1 = getComplex(out, halfLen);

        setComplex(out, 0      , plus (z0, z1));
        setComplex(out, halfLen, minus(z0, z1));
      }
      let rOffset = -rStride;
      for (let k0 = 1, k1 = halfLen + 1; k0 < halfLen; k0++, k1++) {
        const r = rotation(rOffset); rOffset -= rStride;

        const z0 =       getComplex(out, k0);
        const z1 = times(getComplex(out, k1), r);

        setComplex(out, k0, plus (z0, z1));
        setComplex(out, k1, minus(z0, z1));
      }
    }
  }
}

File: fft-rust/rust/src/fft47.rs
use num_complex::{Complex, Complex64};
use std::f64::consts::TAU;

pub struct FFT {
  size: usize,
  cosines: Vec<f64>,
  permute: Vec<usize>,
}

fn rot90(z: Complex64) -> Complex64 {
  Complex{re: -z.im, im: z.re}
}

impl FFT {
  pub fn prepare(size: usize) -> FFT {
    let mut cosines = vec![0.0; size];
    for i in 0..size {
      cosines[i] = (TAU * (i as f64) / (size as f64)).cos();
    }

    let quarter_size = size >> 2;
    let mut permute = vec![0 as usize; quarter_size];
    let mut len = quarter_size;
    let mut f_stride = 1;
    while len > 1 {
      let half_len = len >> 1;
      let mut out_offset: usize = 0;
      while out_offset < quarter_size {
        for out_offset_odd in (out_offset + half_len)..(out_offset + len) {
          permute[out_offset_odd] += f_stride;
        }
        out_offset += len;
      }
      len >>= 1; f_stride <<= 1;
    }

    FFT{size, cosines, permute}
  }

  pub fn run(&self, input: &[Complex64], output: &mut[Complex64], direction: i32) {
    // let FFT{
    //   size: size,
    //   cosines: cosines,
    //   permute: permute,
    // } = self;
    let size = self.size;
    let cosines = &self.cosines;
    let permute = &self.permute;

    // fallback for size <= 2
    match size {
      1 => {
        output[0] = input[0]; return;
      }
      2 => {
        let c0 = input[0];
        let c1 = input[1];
        output[0] = c0 + c1;
        output[1] = c0 - c1;
        return;
        }
      _ => {}
    }

    let mask = size - 1;
    let quarter_size = size >> 2;
    let neg_direction = -direction as f64;

    let rotation = |x| Complex::new(
      cosines[x as usize & mask],
      cosines[(quarter_size - x as usize) & mask]
    );

    let mut out_offset: usize = 0;
    while out_offset < size {
      let mut offset = permute[out_offset >> 2];
      let b0 = input[offset]; offset += quarter_size;
      let b2 = input[offset]; offset += quarter_size;
      let b1 = input[offset]; offset += quarter_size;
      let b3 = input[offset];

      let c0 = b0 + b1;
      let c1 = b0 - b1;
      let c2 = b2 + b3;
      let c3 = rot90(b2 - b3) * neg_direction;

      output[out_offset] = c0 + c2; out_offset += 1;
      output[out_offset] = c1 + c3; out_offset += 1;
      output[out_offset] = c0 - c2; out_offset += 1;
      output[out_offset] = c1 - c3; out_offset += 1;
    }

    let mut len: usize = 8;
    let mut r_stride = direction * (size as i32 >> 3);
    while len < size {
      let half_len = len >> 1;
      // We have pulled out and simplified the case k = 0.
      // TODO Also pull out the case k = quarterLen?
      // r1, r2, and r3 will be -1/8, -2/8, and -3/8 of a full turn, which
      // allows to simplify the expressions for b1, b2, and b3.
      // (But it will not get as simple as the case k = 0.  So I am not sure
      // if it is worthwhile.)
      {
        let mut out_offset = 0;
        while out_offset < size {
          let i0 = out_offset; out_offset += half_len;
          let i1 = out_offset; out_offset += half_len;
          let i2 = out_offset; out_offset += half_len;
          let i3 = out_offset; out_offset += half_len;

          let b0 = output[i0];
          let b1 = output[i1];
          let b2 = output[i2];
          let b3 = output[i3];

          let c0 = b0 + b1;
          let c1 = b0 - b1;
          let c2 = b2 + b3;
          let c3 = rot90(b2 - b3) * neg_direction;
    
          output[i0] = c0 + c2;
          output[i1] = c1 + c3;
          output[i2] = c0 - c2;
          output[i3] = c1 - c3;
        }
      }

      let r_stride1 = r_stride >> 1;
      let r_stride2 = r_stride;
      let r_stride3 = r_stride2 + r_stride1;
      let mut r_offset1 = -r_stride1;
      let mut r_offset2 = -r_stride2;
      let mut r_offset3 = -r_stride3;
      for k in 1..half_len {
        // TODO Some bit fiddling with rOffset[123] to restrict cosine lookups
        // to the first quadrant?  Then shorten the cosines array.
        let r1 = rotation(r_offset1); r_offset1 -= r_stride1;
        let r2 = rotation(r_offset2); r_offset2 -= r_stride2;
        let r3 = rotation(r_offset3); r_offset3 -= r_stride3;
        let mut out_offset = k;
        while out_offset < size {
          let i0 = out_offset; out_offset += half_len;
          let i1 = out_offset; out_offset += half_len;
          let i2 = out_offset; out_offset += half_len;
          let i3 = out_offset; out_offset += half_len;

          let b0 = output[i0];
          let b1 = output[i1] * r2;
          let b2 = output[i2] * r1;
          let b3 = output[i3] * r3;

          let c0 =       b0 + b1;
          let c1 =       b0 - b1;
          let c2 =       b2 + b3;
          let c3 = rot90(b2 - b3) * neg_direction;

          output[i0] = c0 + c2;
          output[i1] = c1 + c3;
          output[i2] = c0 - c2;
          output[i3] = c1 - c3
        }
      }
      len <<= 2; r_stride >>= 2;
    }
    if len == size {
      // If we come here, n is not a power of 4 (but still a power of 2).
      // So we need to run one extra round of 2-way butterflies.
      let half_len = len >> 1;

      // TODO Roll this back into the following loop?
      // Saving a single complex multiplicatin is probably not worth the extra code.
      {
        let z0 = output[0      ];
        let z1 = output[half_len];

        output[0      ] = z0 + z1;
        output[half_len] = z0 - z1;
      }
      let mut r_offset = -r_stride;
      for k0 in 1..half_len {
        let k1 = k0 + half_len;
        let r = rotation(r_offset); r_offset -= r_stride;

        let z0 = output[k0];
        let z1 = output[k1] * r;

        output[k0] = z0 + z1;
        output[k1] = z0 - z1;
      }
    }
  }
}
